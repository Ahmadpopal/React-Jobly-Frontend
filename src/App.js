import React, {useState, useEffect } from 'react';
import { BrowserRouter } from 'react-router-dom'
import  jwt  from 'jsonwebtoken';

import useLocalStorage from './hooks/useLocalStorage'
import UserContext from './Context';
import Routes from './Routes'
import JoblyApi from './Jobly';
import NavbarComponent from './NavbarComponent';



// APP COMPONENT RENDER NAVBAR AND HOLD ROUTES TO THE COMPONENTS

// APP ------> Routes 
// &
// APP -------> Navbar 

function App() {
// INFOLOADED: STATE CHECKS HAS DATA BEEN PULLED FROM JOBLY API 
//     - AND RETURN LOADING MESSAGE FOR DATA UNTIL DATA IS PULLED FROM THE API 
  const [infoLoaded, setInfoLoaded] = useState(false);
  const [applicationIds, setApplicationIds] = useState(new Set([]));
  // currentUser: is user Object from API.
  //   - this is used to check if user is logged In or Not 
  const [currentUser, setCurrentUser] = useState(null);
  // token: is for logged In users, this is their auth JasonWebToken, 
  //   - JWT is decoded in useLocalStorage Hook. 
  const [token, setToken] = useLocalStorage("jobly-token");


   // Load user info from API. Until a user is logged in and they have a token,
  // this should not run. It only needs to re-run when a user logs out, so
  // the value of the token is a dependency for this effect.

  useEffect(function loadUserInfo() {
    console.debug("App useEffect loadUserInfo", "token=", token);

    async function getCurrentUser() {
      if (token) {
        try {
          let { username } = jwt.decode(token);
          // put the token on the Api class so it can use it to call the API.
          JoblyApi.token = token;
          let currentUser = await JoblyApi.getCurrentUser(username);
          setCurrentUser(currentUser);
          setApplicationIds(new Set(currentUser.applications));
        } catch (err) {
          console.error("App loadUserInfo: problem loading", err);
          setCurrentUser(null);
        }
      }
      setInfoLoaded(true);
    }
    // set infoLoaded to false while async getCurrentUser runs; once the
    // data is fetched (or even if an error happens!), this will be set back
    // to false to control the spinner.
    setInfoLoaded(false);
    getCurrentUser();
  }, [token]);



  // logout : Handles user log out in the application 
  function logout() {
    setCurrentUser(null);
    setToken(null);
  }


// register: sign-up or register users to the website 
//   - Async function 
//   - user data is passed to the Api to register 
//     - if successfull will provide a token 
//     other wise catch the error 
  async function register(data){
    try{
      let token = await JoblyApi.signup(data);
      setToken(token);
      return { success: true };
    }catch(errors){
      console.error('signup failed', errors);
      return {success: false, errors };
    }
  }



// login: handle user log accross the website 
//   - async function
//   - provides a token after user log in data is passed correctly 
//   - if errors catch the errors  

  async function logIn(data){
    try{
      let token = await JoblyApi.login(data);
      setToken(token);
      return {success: true};
    } catch(errors){
      console.error("login failed", errors);
      return { success: false, errors };
    }
  }

   /** Checks if a job has been applied for. */
   function hasAppliedToJob(id) {
    return applicationIds.has(id);
  }

  /** Apply to a job: make API call and update set of application IDs. */
  function applyToJob(id) {
    if (hasAppliedToJob(id)) return;
    JoblyApi.applyToJob(currentUser.username, id);
    setApplicationIds(new Set([...applicationIds, id]));
  }
 
  // infoLoaded: if data is loading returns Loading in process 
  if (!infoLoaded) return <h3>Loading...</h3>;


  // BrowserRouter Component: wraps Routes component
  //   - in order to manage routes to the components 
  // UserContext :  manages Props to pass to all Component with passing them nested 
  return (
    <div className="App">
      <BrowserRouter>
      <UserContext.Provider value={{currentUser, setCurrentUser, hasAppliedToJob, applyToJob}}>
        <NavbarComponent  logout={logout}/>
        <Routes login={logIn} register={register} />      
      </UserContext.Provider>
      </BrowserRouter>
    </div>
  );
}

export default App;
