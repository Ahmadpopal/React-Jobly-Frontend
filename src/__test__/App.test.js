import React from 'react'
import { render, screen } from '@testing-library/react';
import App from '../App';

// SMOKE TEST 
test('Component Render without crashing', () => {
  render(<App />)
})


// SNAPSHOT TEST 
test('it matches the snapshot', () =>{
  const { asFragment } = render(<App />)
  expect(asFragment()).toMatchSnapshot()
})

