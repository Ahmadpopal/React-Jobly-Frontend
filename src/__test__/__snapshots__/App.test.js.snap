// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      <nav
        class="navbar bg-dark"
        extend="md"
      >
        <ul
          class="nav"
        >
          <a
            class="nav-link"
            href="/"
          >
            <li
              class="nav-item"
            >
              Jobly
            </li>
          </a>
          <a
            class="nav-link"
            href="/login"
          >
            <li
              class="nav-item"
            >
              Log In
            </li>
          </a>
          <a
            class="nav-link"
            href="/register"
          >
            <li
              class="nav-item"
            >
              Sign Up
            </li>
          </a>
        </ul>
      </nav>
    </div>
    <div>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="HomePage-main-div col"
          >
            <div
              class="row"
            >
              <div
                class="HomePage-div-title col"
              >
                <h3>
                  Welcome to Jobly Home Page.
                </h3>
                <h3>
                  Open jobs.
                </h3>
                <h3>
                  People hiring.
                </h3>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <button
                  class="HomePage-login-button btn btn-success"
                >
                  <a
                    class="HomePage-login-link"
                    href="/login"
                  >
                    Log In
                  </a>
                </button>
                <button
                  class="HomePage-signup-button btn btn-primary"
                >
                  <a
                    class="HomePage-signup-button"
                    href="/register"
                  >
                    Sign Up
                  </a>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
